# Pipeline citations

## [Snakemake](https://snakemake.readthedocs.io/en/stable/index.html)

> Mölder, F., Jablonski, K.P., Letcher, B. et al. Sustainable data analysis with Snakemake [version 2; peer review: 2 approved]. F1000Research, 10(33), (2021). https://doi.org/10.12688/f1000research.29032.2

## Pipeline tools

**`ABRicate`**

> Seemann, T., ABRicate, (2024), Github https://github.com/tseemann/abricate

**`CheckM2`**

> Chklovski, A., Parks, D.H., Woodcroft, B.J. et al. CheckM2: a rapid, scalable and accurate tool for assessing microbial genome quality using machine learning. Nature Methods, 20, (2023). https://doi.org/10.1038/s41592-023-01940-w

**`CoverM`**

> Aroney, S.T., Newell, R.J., Nissen, J.N., Camargo, A.P., Tyson, G.W. and Woodcroft, B.J., CoverM: Read alignment statistics for metagenomics. arXiv preprint arXiv:2501.11217, (2025). https://doi.org/10.48550/arXiv.2501.11217

**`emmtyper`**

> Tan, A., Seemann, T., Lacey, J., Davies, M., Mcintyre, L., Frost, H., Wiliamson, D., Gonçalves da Silva, A. emmtyper, (2025), Github https://github.com/MDU-PHL/emmtyper

**`Flye`**

> Kolmogorov, M., Bickhart, D.M., Behsaz, B. et al. metaFlye: scalable long-read metagenome assembly using repeat graphs. Nature Methods, 17, (2020). https://doi.org/10.1038/s41592-020-00971-x

**`KMA`**

> Clausen, P.T.L.C., Aarestrup, F.M. & Lund, O. Rapid and precise alignment of raw reads against redundant databases with KMA. BMC Bioinformatics, 19(307), (2018). https://doi.org/10.1186/s12859-018-2336-6

**`KmerFinder`**

> Larsen MV, Cosentino S, Lukjancenko O, Saputra D, Rasmussen S, Hasman H, Sicheritz-Pontén T, Aarestrup FM, Ussery DW, Lund O. Benchmarking of Methods for Genomic Taxonomy. Journal of Clinical Microbiology, 52, (2014). https://doi.org/10.1128/jcm.02981-13

**`KmerResistance`**

> Clausen, P.T.L.C., Zankari, E., Aarestrup, F.M., Lund, O. Benchmarking of methods for identification of antimicrobial resistance genes in bacterial whole genome data. Journal of Antimicrobial Chemotherapy, 71(9), (2016). https://doi.org/10.1093/jac/dkw184 and Larsen, M.V., Cosentino, S., Lukjancenko, O., Saputra, D., Rasmussen, S., Hasman, H., Sicheritz-Pontén,T ., Aarestrup, F.M., Ussery, D.W., Lund, O. 2014. Benchmarking of Methods for Genomic Taxonomy. J Clin Microbiol 52:https://doi.org/10.1128/jcm.02981-13

**`Kraken2`**

> Wood, D.E., Lu, J., Langmead, B. Improved metagenomic analysis with Kraken 2. Genome Biology, 20(257), (2019). https://doi.org/10.1186/s13059-019-1891-0

**`NanoPlot`**

> De Coster, W., Rademakers, R. NanoPack2: population-scale evaluation of long-read sequencing data. Bioinformatics, 39(5), (2023). https://doi.org/10.1093/bioinformatics/btad311

**`NanoQ`**

> Steinig, E., Coin, L.J.M. Nanoq: ultra-fast quality control for nanopore reads. Journal of Open Source Software, 7(69), (2022). https://doi.org/10.21105/joss.02991

**`NoHuman`**

> Hall, M.B., Coin, L.J.M. Pangenome databases improve host removal and mycobacteria classification from clinical metagenomic data. GigaScience, 13, (2024). https://doi.org/10.1093/gigascience/giae010

**`R`**

> R Core Team. R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. (2025). URL https://www.R-project.org/.

**`ResFinder`**

> Zankari, E., Hasman, H., Cosentino, S., Vestergaard, M., Rasmussen, S., Lund, O., Aarestrup, F.M., Larsen, M.V. Identification of acquired antimicrobial resistance genes. Journal of Antimicrobial Chemotherapy, 67(11), (2012). https://doi.org/10.1093/jac/dks261 

**`StarAMR`**

> Bharat, A., Petkau, A., Avery, B. P., Chen, J. C., Folster, J. P., Carson, C. A., Kearney, A., Nadon, C., Mabon, P., Thiessen, J., Alexander, D. C., Allen, V., El Bailey, S., Bekal, S., German, G. J., Haldane, D., Hoang, L., Chui, L., Minion, J., Zahariadis, G., Van Domselaar, G., Reid-Smith, R.J., Mulvey, M. R. Correlation between Phenotypic and In Silico Detection of Antimicrobial Resistance in Salmonella enterica in Canada Using Staramr. Microorganisms, 10(2), (2022). https://doi.org/10.3390/microorganisms10020292

**`Sylph`**

> Shaw, J., Yu, Y.W. Rapid species-level metagenome profiling and containment estimation with sylph. Nature Biotechnology, (2024). https://doi.org/10.1038/s41587-024-02412-y

**`VFDB`**

> Chen, L., Zheng, D., Liu, B., Yang, J., Jin, Q. VFDB 2016: hierarchical and refined dataset for big data analysis—10 years on, Nucleic Acids Research, 44(D1), (2016). https://doi.org/10.1093/nar/gkv1239 
